Improve the flow for adding a question to a dashboard

calherries opened the pull request:
Feature branch for epic [Improve the flow for adding a question to a dashboard](#29666)

### Description
To improve UX of adding question to the dashboard, we added next changes:

- if there was a dashboard visited during last 24hrs, we preselect this dashboard in the ItemPicker by default
- if there wasn't any dashboard visited during last 24hrs, we show root collection

### How to verify

- Go to existing or create a new question
- Click Add to dashboard
- make sure the dashboard you visited recently is preselected in the picker

### Demo

https://www.loom.com/share/41ba6fbd4a2e44b099e9a05206db85a3

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/metabase/metabase/30347)
<!-- Reviewable:end -->

calherries committed:
Add GET "/most_recently_viewed_dashboard" endpoint (#30323)

* Only update dashboard if changes have happened

* Add GET "/most_recently_viewed_dashboard" endpoint

* Add endpoint

* Undo accidentally committed changes to api/dashboard/:id

* Undo accidental changes to record-view!

* Fix test

* Fix test docstring

uladzimirdev committed:
Merge remote-tracking branch 'origin/master' into question-to-dash-flow/main

calherries committed:
Return dashboard in GET "/most_recently_viewed_dashboard" response (#30566)

* Only update dashboard if changes have happened

* Add GET "/most_recently_viewed_dashboard" endpoint

* Add endpoint

* Undo accidentally committed changes to api/dashboard/:id

* Undo accidental changes to record-view!

* Fix test

* Fix test docstring

* Return dashboard object and add permissions check

* Return 204 if no recent viewed dashboard

Three cases to handle:
- no recent dashboard
- recent dashboard but lack permissions
- recent dashboard you can see

the FE really just cares "is there a recent dashboard to show you" which
logically collapses the first two cases. So we return a generic 204
to the FE. In the third case we return the dashboard.

* Fixes IP geocoding test (#30605)

* Typo

---------

Co-authored-by: dan sutton <dan@dpsutton.com>
Co-authored-by: john-metabase <92878045+john-metabase@users.noreply.github.com>
Co-authored-by: Uladzimir Havenchyk <uladzimir.dev@gmail.com>

uladzimirdev committed:
Merge remote-tracking branch 'origin/master' into question-to-dash-flow/main

uladzimirdev committed:
Merge remote-tracking branch 'origin/master' into question-to-dash-flow/main

uladzimirdev committed:
Preselect most recently viewed dashboard in DashboardPicker (#30581)

* Only update dashboard if changes have happened

* Add GET "/most_recently_viewed_dashboard" endpoint

* Add endpoint

* Undo accidentally committed changes to api/dashboard/:id

* Undo accidental changes to record-view!

* Fix test

* Fix test docstring

* Return dashboard object and add permissions check

* Preselect most recently viewed dashboard in DashboardPicker

* Use correct type and fix loading logic

* Handle case when no dashboard was visited during last 24hrs

* Return 204 if no recent viewed dashboard

Three cases to handle:
- no recent dashboard
- recent dashboard but lack permissions
- recent dashboard you can see

the FE really just cares "is there a recent dashboard to show you" which
logically collapses the first two cases. So we return a generic 204
to the FE. In the third case we return the dashboard.

* Fixes IP geocoding test (#30605)

* Typo

* WIP

* Add test to AddToDashSelectDashModal

* fixup! Add test to AddToDashSelectDashModal

* Handle api errors in entity hooks (#30687)

* Address review commenents on test

* Reorder tests

* Fixup

* Add another test

* prettier

* Fixup

* Fixup

* Address edge case

* Move helpers to server-mocks

* Move helpers to server-mocks

* Add collection query hook

---------

Co-authored-by: Callum Herries <hi@callumherries.com>
Co-authored-by: Cal Herries <39073188+calherries@users.noreply.github.com>
Co-authored-by: dan sutton <dan@dpsutton.com>
Co-authored-by: john-metabase <92878045+john-metabase@users.noreply.github.com>
Co-authored-by: Alexander Polyankin <alexander.polyankin@metabase.com>

uladzimirdev committed:
Merge branch 'master' into question-to-dash-flow/main

mazameli reviewed (approved):
This is great! I tried these two flows:
1. New > Question > Save the question and follow the dialog prompt to add it to a dashboard.
2. Existing saved question > Add To Dashboard button

cdeweyx commented:
Much smoother, nice work 🙌

uladzimirdev committed:
Cleanup server-mocks (#30811)

* Cleanup server-mocks

* Handle personal collection

* Rework and simplify test helper

* Handle special case in picker item

When fake collection with id "personal" is used, we don't need to make http request, only several personal collections can be listed

* Fixup
